include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContributionApproximator.h

```diff
@@ -0,0 +1,397 @@
+
+			//section 3.4.1: push the least contributor: decrease its delta further to have a chance to end earlier.
+			if( activePoints.size() > 2 ) {
+				sample( **minimalElement, round, m_minimumMultiplierDelta * delta, n );
+				minApprox = std::numeric_limits<double>::max();
+				for( auto it = activePoints.begin(); it != activePoints.end(); ++it ) {
+					if( (*it)->approximatedContribution < minApprox ) {
+						minApprox = (*it)->approximatedContribution;
+						minimalElement = it;
+					}
+				}
+			}
+
+			//remove all points whose confidence interval does not overlap with the current minimum any more.
+			auto it = activePoints.begin();
+			while( it != activePoints.end() ) {
+				if( (*minimalElement)->contributionUpperBound  < (*it)->contributionLowerBound )
+					it = activePoints.erase( it );
+				else
+					++it;
+			}
+
+			// stopping conditions: have we reached the desired accuracy? 
+			// for this we need to know:
+			// 1. contribution for all points are bounded above 0
+			// 2. upperBound(LC) < (1+epsilon)*lowerBound(A) for all A
+			double d = 0;
+			for( it = activePoints.begin(); it != activePoints.end(); ++it ) {
+				if( it == minimalElement )
+					continue;
+				double nom = (*minimalElement)-> contributionUpperBound;
+				double den = (*it)->contributionLowerBound;
+				if( den <= 0. )
+					d = std::numeric_limits<double>::max();
+				else
+					d = std::max(d,nom/den);
+			}
+			
+			//if the stopping condition is fulfilled, return the minimal element
+			if(d < 1. + m_errorBound){
+				return *minimalElement;
+			}
+			
+			delta *= m_multiplierDelta;
+		}
+	}
+	
+	/// \brief Samples in the bounding box of the supplied point until a pre-defined threshold is reached.
+	///
+	/// \param [in] s Set of points.
+	/// \param [in] point Iterator to the point that should be sampled.
+	/// \param [in] r The current round.
+	/// \param [in] delta The delta that should be reached.
+	template<class VectorType>
+	void sample( Point<VectorType>& point, unsigned int r, double delta, std::size_t n )const{
+		double logFactor = std::log( 2. * n * (1. + m_gamma) / (m_errorProbability * m_gamma) );
+		double logR = std::log( static_cast<double>( r ) );
+		//compute how many samples we need until the bound of the current box is smaller than delta
+		//this is formula (3) in the paper when used in an equality < delta and solving for noSamples
+		//we add +1 to ensure that the inequality holds.
+		double thresholdD= 1.0+0.5 * ( (1. + m_gamma) * logR + logFactor ) * sqr( point.boundingBoxVolume / delta );
+		std::size_t threshold = static_cast<std::size_t>(thresholdD);
+		//sample points inside the box of the current point
+		for( ; point.noSamples < threshold; point.noSamples++ ) {
+			//sample a point inside the box
+			point.sample.resize(point.point.size());
+			for( unsigned int i = 0; i < point.sample.size(); i++ ) {
+				point.sample[ i ] = point.point[ i ] + Rng::uni( 0., 1. ) * ( point.boundingBox[ i ] - point.point[ i ] );
```

Comment:
I think that this line should read

    point.sample[ i ] = Rng::uni(point.point[ i ], point.boundingBox[ i ]);